libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp

```diff
@@ -0,0 +1,759 @@
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* Protocol implementation was provided by FETtec */
+
+#include <AP_Math/AP_Math.h>
+#include <AP_Scheduler/AP_Scheduler.h>
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <SRV_Channel/SRV_Channel.h>
+#include <GCS_MAVLink/GCS.h>
+#include <AP_Arming/AP_Arming.h>
+
+#include "AP_FETtecOneWire.h"
+#if HAL_AP_FETTEC_ONEWIRE_ENABLED
+
+static constexpr uint32_t FTOW_UART_LOCK_KEY = 0x0FE77EC0;
+static constexpr uint32_t DELAY_TIME_US = 700;
+static constexpr uint32_t RESPONSE_LENGTH = 18;
+static constexpr uint8_t ALL_ID = 0x1F;
+static constexpr uint8_t MAX_TRANSMIT_LENGTH = 4;
+static constexpr uint8_t MAX_RECEIVE_LENGTH = 14;
+
+const AP_Param::GroupInfo AP_FETtecOneWire::var_info[] = {
+    // @Param: MASK
+    // @DisplayName: Channel Bitmask
+    // @Description: Enable of FETtec OneWire ESC protocol to specific channels
+    // @Bitmask: 0:Channel1,1:Channel2,2:Channel3,3:Channel4,4:Channel5,5:Channel6,6:Channel7,7:Channel8,8:Channel9,9:Channel10,10:Channel11,11:Channel12,12:Channel13,13:Channel14,14:Channel15,15:Channel16
+    // @RebootRequired: True
+    // @User: Standard
+    AP_GROUPINFO("MASK",  1, AP_FETtecOneWire, motor_mask, 0),
+
+    // @Param: POLES
+    // @DisplayName: Nr. electrical poles
+    // @Description: Number of motor electrical poles
+    // @Range: 2 50
+    // @RebootRequired: False
+    // @User: Standard
+    AP_GROUPINFO("POLES", 2, AP_FETtecOneWire, pole_count, 14),
+
+    AP_GROUPEND
+};
+
+AP_FETtecOneWire *AP_FETtecOneWire::_singleton;
+
+AP_FETtecOneWire::AP_FETtecOneWire()
+{
+    _singleton = this;
+
+    _response_length[OW_OK] = 1;
+    _response_length[OW_BL_START_FW] = 0;       // BL only
+    _response_length[OW_REQ_TYPE] = 1;
+    _response_length[OW_REQ_SN] = 12;
+    _response_length[OW_REQ_SW_VER] = 2;
+    _response_length[OW_SET_FAST_COM_LENGTH] = 1;
+    _response_length[OW_SET_TLM_TYPE] = 1;
+
+    _request_length[OW_OK] = 1;
+    _request_length[OW_BL_START_FW] = 1;       // BL only
+    _request_length[OW_REQ_TYPE] = 1;
+    _request_length[OW_REQ_SN] = 1;
+    _request_length[OW_REQ_SW_VER] = 1;
+    _request_length[OW_SET_FAST_COM_LENGTH] = 4;
+    _request_length[OW_SET_TLM_TYPE] = 2;
+}
+
+void AP_FETtecOneWire::init()
+{
+    AP_SerialManager& serial_manager = AP::serialmanager();
+    _uart = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FETtecOneWire, 0);
+    if (_uart) {
+        _uart->set_flow_control(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE);
+        _uart->set_unbuffered_writes(true);
+        _uart->set_blocking_writes(false);
+        _uart->begin(500000);
+        _initialised = true;
+    }
+}
+
+bool AP_FETtecOneWire::configuration_ok()
+{
+    if (_update_loop_decimator-- == 0) {
+        _update_loop_decimator = 2 * _update_rate_hz; // once every 2 seconds

```

Comment:
it's much cleaner to decimate based on a timestamp - then the hz is fixed, no matter if the update rate changes

Reply:
This is supposed to be as lightweight as possible. Requires no function calls at all.

Reply:
Pretty sure that if a call to AP_HAL::millis() is expensive then we are all in trouble ...

Reply:
I did it now